src/org/rascalmpl/repl/TerminalProgressBarMonitor.java

```diff
@@ -0,0 +1,459 @@
+
+            // capitalize
+            var msg = message.substring(0, 1).toUpperCase() + message.substring(1, message.length());
+            
+            // fill up and cut off:
+            msg = threadLabel() + msg;
+            msg = (msg + " ".repeat(Math.max(0, barWidth - msg.length()))).substring(0, barWidth);
+
+            // split
+            var frontPart = msg.substring(0, doneWidth);
+            var backPart = msg.substring(doneWidth, msg.length());
+            var clock = clocks[stepper % clocks.length];
+
+            if (barWidth < 1) {
+                return; // robustness against very small screens. At least don't throw bounds exceptions
+            }
+            else if (barWidth <= 3) { // we can print the clock for good measure 
+                writer.println(clock);
+                return;
+            }
+
+            var line 
+                = done 
+                + ANSI.darkBackground() 
+                + frontPart
+                + ANSI.noBackground()
+                + ANSI.lightBackground()
+                + backPart
+                + ANSI.noBackground()
+                + " " + clock + " "
+                + String.format("%d:%02d:%02d.%03d", duration.toHoursPart(), duration.toMinutes(), duration.toSecondsPart(), duration.toMillisPart())
+                + " "
+                ;
+
+            writer.println(line);
+        }
+
+        private String threadLabel() {
+            if (threadName.isEmpty()) {
+                return "";
+            }
+            else if ("main".equals(threadName)) {
+                return "";
+            }
+            
+            return threadName + ": ";
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            return obj instanceof ProgressBar 
+                && ((ProgressBar) obj).name.equals(name)
+                && ((ProgressBar) obj).threadId == threadId
+                ;
+        }
+
+        @Override
+        public int hashCode() {
+            return name.hashCode() + 31 * (int) threadId;
+        }
+
+        public void done() {
+            this.current = Math.min(current, max);
+            this.duration = Duration.between(startTime, Instant.now());
+            this.message = name;
+        }
+    }
+
+    /**
+     * Clean the screen, ready for the next output (normal output or the next progress bars or both).
+     * We use fancy ANSI codes here to move the cursor and clean the screen to remove previous versions
+     * of the bars
+     */
+    private void eraseBars() {
+        if (bars.isEmpty()) {
+            // writer.println();
+            return;
+        }
+            
+        writer.write(ANSI.hideCursor());
+        writer.write(ANSI.moveUp(bars.size())); 
+        writer.write(ANSI.clearToEndOfScreen()); 
+        writer.write(ANSI.showCursor());
+        writer.flush();
+    }
+
+    /**
+     * ANSI escape codes convenience functions
+     */
+    private static class ANSI {
```

Comment:
Is there a reason we're not using ANSI class from jline? We have used that in other parts of the REPL.